### PR TITLE
monitoring-satellite: Replace all extVars with centralized variable

### LINKED ISSUE #34: Find a way to introduce new extVars in a backwards compatible way
jsonnet [deliberately doesn't support default values for external variables](https://github.com/google/go-jsonnet/issues/247#issuecomment-425662159) which means that whenever we introduce a new `extVar` we will break all existing invocations. So at the moment you have to introduce the `extVar` everywhere (jsonnet will happily ignore unused `extVar`s) before you start using it in your templates.

Right now that's annoying but tolerable as we own all invocations of our jsonnet templates. But soon (maybe already) that isn't the case as the Workspace team will start consuming them.

We have a few options (currently we favour option 3)

1. Accept that introducing extVars is a breaking change. All users of our templates should then point to a fixed SHA of this repository and we can maintain a CHANGELOG that consumers can look at when upgrading.
2. Introduce an indirection: We publish a shell script (as the repo is public people can just use the Git url) which takes all extVars as parameters and can thus pick default values and then invoke our jsonnnet templates with all extVars set.
3. **Use a json configuration file (later it can be YAML once jsonnet supports it) that our jsonnet programs loads. That file becomes the configuration surface and our jsonnet programs can provide default values as needed.**

### PR DESCRIPTION
**Pull request by ArthurSens**
## Description

I'm opening this PR to share what I've been experimenting with extVars!

The idea is to replace all variables with a single one, which should contain a JSON with whatever needs to be configured.

The main benefit here is that we can now use `std.ObjectHas(object, field)` to verify if a certain piece of configuration was added or not, instead of asking for a boolean variable like `alerting_enabled`. When running jsonnet with:

```
jsonnet --ext-code config="{
    namespace: 'monitoring-satellite',
    clusterName: 'fake-cluster',
    alerting: {
        slackWebhookURLCritical: 'http://fake.url.critical',
        slackWebhookURLWarning: 'http://fake.url.warning',
        slackWebhookURLInfo: 'http://fake.url.info',
        slackChannelPrefix: '#fake_channel',
        pagerdutyRoutingKey: 'fakeR0uT1GNKEY',
    },
}"
```
In code something like this

```jsonnet
(if std.extVar('alerting_enabled') == 'true' then (import '../addons/alerting.libsonnet') else {}) 
(if std.extVar('tracing_enabled') == 'true' then (import '../addons/alerting.libsonnet') else {}) 
```
Turns into:

```jsonnet
(if std.objectHas(config, 'alerting') then (import '../addons/alerting.libsonnet') else {}) 
(if std.objectHas(config, 'tracing') then (import '../addons/alerting.libsonnet') else {}) 
```

The alerting addon would be added while tracing would not.
We can add new features under new fields without requiring new external variables to be set.

---

## Possible pitfalls

This strategy does allow extra flexibility for adding new features but it still isn't 100% bullet-proof against breaking changes! Once we agree on a JSON Scheme for our configuration and announce it to the engineering teams, changing the scheme would be a breaking change.

This strategy only allows us to add new fields to our config scheme, which is already better than what we have today 🙂

### CODE REVIEW

**Comment by ArthurSens** [comment]
Check out the changes made to `hack/generate.sh` for an example of what an initial configuration scheme would look like!

File: hack/generate.sh [hunk 1]
```diff
@@ -94,21 +105,11 @@ monitoring-central/manifests/yaml-generator.jsonnet | xargs -I{} sh -c 'cat {} |
...
 # Generate monitoring-satellite prometheus rules
 jsonnet -c -J vendor -m monitoring-satellite/manifests \
---ext-str namespace="monitoring-satellite" \
---ext-str cluster_name="fake-cluster" \
---ext-str slack_webhook_url_critical="http://fake.url.critical" \
---ext-str slack_webhook_url_warning="http://fake.url.warning" \
---ext-str slack_webhook_url_info="http://fake.url.info" \
---ext-str slack_channel_prefix="#fake_channel" \
---ext-str pagerduty_routing_key="fakeR0uT1GNKEY" \
 --ext-str node_affinity_label='' \
```
Comment (mads-hartmann) [inline hack/generate.sh]: You can remove this `--ext-str` as well right? :)

**Review by mads-hartmann — COMMENTED** [review]
This looks like a really promising way forward to me.

We could even extend the templates to do deep validation of the configuration right? E:g. we could have a `validate-config.libsonnet` that we would import which would then contain a bunch of `assert`s like the following (consider this pseudo jsonnet, I didn't run this):

```jsonnet
function validate_preview_environment_config(config) =
  assert std.objectHas(config.previewEnvironment, 'nodeExporterPort') : "Missing required field: previewEnvironment.nodeExporterPort" ;
  assert .... 

assert std.objectHas(config, 'clusterName') : "Missing cluserName key in configuration";

if std.objectHas(config, 'previewEnvironment') then validate_preview_environment_config(config)
// and so on
```

**Comment by ArthurSens** [comment]
> This looks like a really promising way forward to me.
> 
> We could even extend the templates to do deep validation of the configuration right? E:g. we could have a `validate-config.libsonnet` that we would import which would then contain a bunch of `assert`s like the following (consider this pseudo jsonnet, I didn't run this):
> 
> ```jsonnet
> function validate_preview_environment_config(config) =
>   assert std.objectHas(config.previewEnvironment, 'nodeExporterPort') : "Missing required field: previewEnvironment.nodeExporterPort" ;
>   assert .... 
> 
> assert std.objectHas(config, 'clusterName') : "Missing cluserName key in configuration";
> 
> if std.objectHas(config, 'previewEnvironment') then validate_preview_environment_config(config)
> // and so on
> ```

Definitely, that's something that we could do! Another thing that I'd like to get rid of is that I'm using the same `config` extVar multiple times across some addons and libraries under `components/`. It would be nice if we could figure out a nice way of importing this config only once and pass it to all other places as functions+arguments

**Review by meysholdt — APPROVED** [review]
code changes LGTM :)

**Comment by ArthurSens** [comment]
> code changes LGTM :)

I believe there is still a lot to discuss here 😬. This is a draft PR just to share some possibilities...

One thing that we definitely need to do before merging:
* Agree on a configuration scheme

Things that could be left to follow-up PRs(But are still important):
* Find a way to import the extVar(config) in a single place and pass it down through function+arguments (At the moment I'm importing it in several addons and libraries)
* Rewrite monitoring-central to follow this same pattern

**Comment by ArthurSens** [comment]
> Find a way to import the extVar(config) in a single place and pass it down through function+arguments (At the moment I'm importing it in several addons and libraries)

I think I figured that out in this commit: c5bbe38 🙂 @mads-hartmann @meysholdt what do you think?

**Comment by mads-hartmann** [comment]
@ArthurSens I like the approach in c5bbe38 - means that the config validation is local to the file (so might be easier to maintain and update as you work on a specific file). Having the templates be parameterised by the configuration also means that it will be much easier to write tests if we ever wanted to write [jsonnetunit](https://github.com/yugui/jsonnetunit) tests :)

**Pull request merged**